File: held-karp-cpp/heldkarp.cpp
#include <bits/stdc++.h>
using namespace std;

static const double INF = 1e18;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<vector<double>> dist(n, vector<double>(n));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            cin >> dist[i][j];

    if (n == 1) {
        cout << "0\n0\n";
        return 0;
    }

    int N = n - 1;                 // exclude start node 0
    int FULL = 1 << N;

    vector<double> dp(FULL * n, INF);
    vector<int> parent(FULL * n, -1);

    auto IDX = [&](int mask, int u) { return mask * n + u; };

    // base: from start → each node
    for (int u = 1; u < n; u++) {
        int m = 1 << (u - 1);
        dp[IDX(m, u)] = dist[0][u];
    }

    for (int mask = 1; mask < FULL; mask++) {
        for (int u = 1; u < n; u++) {
            if (!(mask & (1 << (u - 1)))) continue;

            double cur = dp[IDX(mask, u)];
            if (cur >= INF) continue;

            int remaining = ((FULL - 1) ^ mask);

            while (remaining) {
                int bit = remaining & -remaining;
                int v = __builtin_ctz(bit) + 1;
                remaining ^= bit;

                int newMask = mask | (1 << (v - 1));
                double newCost = cur + dist[u][v];

                double &ref = dp[IDX(newMask, v)];
                if (newCost < ref) {
                    ref = newCost;
                    parent[IDX(newMask, v)] = u;
                }
            }
        }
    }

    double best = INF;
    int last = -1;
    int finalMask = FULL - 1;

    for (int u = 1; u < n; u++) {
        double cost = dp[IDX(finalMask, u)] + dist[u][0];
        if (cost < best) {
            best = cost;
            last = u;
        }
    }

    // reconstruct tour
    vector<int> tour = {0};
    vector<int> path;

    int mask = finalMask;
    int cur = last;

    while (cur != -1) {
        path.push_back(cur);
        int p = parent[IDX(mask, cur)];
        mask ^= (1 << (cur - 1));
        cur = p;
    }

    reverse(path.begin(), path.end());

    for (int v : path) tour.push_back(v);
    tour.push_back(0);

    cout << best << "\n";
    for (int v : tour) cout << v << " ";
    cout << "\n";

    return 0;
}

File: lib/solvers/held-karp.ts
/**
 * Held-Karp Algorithm — Exact TSP solver using bitmask dynamic programming.
 *
 * Binary Encoding (Quantum Mapping):
 *   Each subset S of cities is encoded as a bitmask integer where bit i = 1
 *   means city i is included in the subset.
 *
 * DP Recurrence:
 *   dp[S][i] = min cost to visit all cities in subset S, ending at city i
 *   dp[S][i] = min over j in S\{i} of { dp[S\{i}][j] + dist[j][i] }
 *
 * Complexity: O(n² · 2ⁿ)  — feasible for n ≤ 18
 */

export interface SolverResult {
    tour: number[];
    distance: number;
    solverName: string;
    timeMs?: number;
    skipped?: boolean;
    reason?: string;
}

// Must match the value in modifiers.ts — edges with cost >= this are blocked
const BLOCKED_THRESHOLD = 1e9;

export function heldKarp(dist: number[][], startNode: number = 0): SolverResult {
    const n = dist.length;

    if (n > 20) {
        throw new Error(`Held-Karp is infeasible for n=${n} (max 18–20). Use Nearest Neighbor instead.`);
    }

    if (n === 1) {
        return { tour: [startNode], distance: 0, solverName: 'Held-Karp (Exact)' };
    }

    if (n === 2) {
        const other = startNode === 0 ? 1 : 0;
        return {
            tour: [startNode, other, startNode],
            distance: dist[startNode][other] + dist[other][startNode],
            solverName: 'Held-Karp (Exact)',
        };
    }

    const FULL_MASK = (1 << n) - 1;
    // Use a value larger than any possible sum of blocked edges so blocked
    // paths always lose to non-blocked ones. n * BLOCKED_THRESHOLD guarantees
    // that even a full tour of blocked edges sums to less than INF.
    const INF = (n + 1) * BLOCKED_THRESHOLD;

    const dp = Array.from({ length: 1 << n }, () => new Float64Array(n).fill(INF));
    const parent = Array.from({ length: 1 << n }, () => new Int8Array(n).fill(-1));

    // Base case: start at startNode
    dp[1 << startNode][startNode] = 0;

    // Fill DP table
    for (let mask = 1; mask <= FULL_MASK; mask++) {
        if (!(mask & (1 << startNode))) continue;

        for (let u = 0; u < n; u++) {
            if (!(mask & (1 << u))) continue;
            if (dp[mask][u] === INF) continue;

            for (let v = 0; v < n; v++) {
                if (mask & (1 << v)) continue;
                const newMask = mask | (1 << v);
                const newCost = dp[mask][u] + dist[u][v];

                if (newCost < dp[newMask][v]) {
                    dp[newMask][v] = newCost;
                    parent[newMask][v] = u;
                }
            }
        }
    }

    // Find optimal last city before returning to start
    let minCost = INF;
    let lastCity = -1;

    for (let u = 0; u < n; u++) {
        if (u === startNode) continue;
        const totalCost = dp[FULL_MASK][u] + dist[u][startNode];
        if (totalCost < minCost) {
            minCost = totalCost;
            lastCity = u;
        }
    }

    // Reconstruct tour
    const tour: number[] = [startNode];
    let mask = FULL_MASK;
    let current = lastCity;

    const path: number[] = [];
    while (current !== startNode) {
        path.push(current);
        const prev = parent[mask][current];
        mask ^= 1 << current;
        current = prev;
    }

    path.reverse();
    tour.push(...path);
    tour.push(startNode); // Return to depot

    return {
        tour,
        distance: minCost,
        solverName: 'Held-Karp (Exact)',
    };
}
